### PR TITLE
feat(operation): complex Householder, QR and LQ

### PR DESCRIPTION
**Pull request by Ravenwater**
## Summary

**Part 2 of #353.** The cholesky half is #360; this is the householder/qr half. Both use `Relates to`, so #353 closes when the second merges.

`householder()` failed to compile for `std::complex` on `axi > scale` and `x0 <= 0` — relational operators applied to a complex where a **magnitude** was meant. As #353 anticipated, the gap was larger than the comparisons: `sigma` accumulated `z²` rather than `|z|²`, and the reflector was written `I − β·v·vᵀ` rather than `I − τ·v·vᴴ`. Repairing only the types would have compiled and produced a wrong Q.

## The construction

`H = I − τ·v·vᴴ` with `H·x = β·e₁`, β real (LAPACK `zlarfg`). With `v = [1; u]`, `x = [a; y]`:

```
u   = y / (a - beta)
tau = (beta - conj(a)) / beta
```

β's **sign** is the free choice, taken opposite to `Re(a)` so `a − β` cannot cancel — the complex analogue of the real branch's `x0 <= 0` test.

## Why each consumer differs

`H` is unitary but **not Hermitian** for complex τ, so `H⁻¹ = Hᴴ`, not `H`. That one fact is the substance of the change, and it lands differently in each place:

| site | applies | why |
|---|---|---|
| `qr_extract_Q` | `conj(tau)` | `R = H_{k-1}…H_0 A` ⟹ `Q = H_0ᴴ … H_{k-1}ᴴ` |
| `qr_solve` | `tau`, conj on `v` in the inner product | wants `Qᴴb = H_{k-1}…H_0 b` |
| `lq_factor` | reflector from the **conjugated row**, applied as `Hᴴ` | see below |
| `lq_extract_Q` | plain `tau` | the adjoint was already taken on the factor side |

Every one of those is invisible for real input, where `conj` is the identity.

## The LQ case, and why the reconstruction check earns its keep

My first attempt reasoned by analogy with QR — "the inverse of a complex reflector is its adjoint, so use `conj(tau)` in `lq_extract_Q`". It produced a **perfectly unitary Q** while `L·Q` differed from `A` by **O(1)**:

```
LQ m=4 n=4  |Q^H Q - I|=5.55e-16  |LQ - A|=8.77e-01
LQ m=3 n=6  |Q^H Q - I|=4.44e-16  |LQ - A|=9.50e-01
```

**A unitarity check alone would have passed it.** The error is that `householder()` annihilates a **column** while LQ annihilates a **row**, and for complex those are different problems. With `c = rᴴ`:

```
r * H^H = c^H H^H = (H c)^H = beta * e_1^T
```

so the reflector must be built from the conjugated row **and** applied as `Hᴴ` — both halves or neither. The test therefore checks `L·Q = A`, not just `QᴴQ = I`.

## Rejecting what is not fixed

`hessenberg_factor` and `eigenvalue_symmetric_generic` apply `H·A·H`, a similarity transform **only because a real reflector is Hermitian and involutory**. For complex, `H⁻¹ = Hᴴ ≠ H`, so the same code computes a non-similar matrix — different eigenvalues, no diagnostic.

Making `householder()` complex-capable would have made both of them *compile* for complex. **A fix in one file would have opened a silent wrong-answer path in two others.** Both now reject complex explicitly, pinned by compile-failure tests using the #358 harness. The Hermitian reduction (real subdiagonal, phase accumulation) is its own piece of work.

## Verification

Measured, not assumed:

| check | result |
|---|---|
| `H·x = β·e₁` | β imag 4.4e-16; `\|β\| == ‖x‖` to 8.9e-16 |
| `H` unitary | `‖HᴴH − I‖ = 2.2e-16` |
| complex QR — 4×4, 6×3 (tall), 3×6 (wide), 5×5 | `‖QᴴQ−I‖ ≤ 4.4e-16`, `‖QR−A‖ ≤ 7.0e-16` |
| complex LQ — 4×4, 3×6, 6×3 | `‖QᴴQ−I‖ ≤ 4.4e-16`, `‖LQ−A‖ ≤ 6.9e-16` |
| complex `qr_solve` | `max\|x − x_true\| = 3.3e-16` |
| **real path vs `main`** | **bit-identical** |

The real-path check is a hex-float digest of `tau`, `Q(0,0)`, `Q(3,4)` and the full solve vector, compared against the same program compiled on stashed (main) code — not a tolerance. The real branch of `householder()` is kept **verbatim** inside an `if constexpr` precisely so this holds; folding both cases into one general formulation would have perturbed existing results for no benefit.

## Changes

- `householder.hpp` — magnitudes in `magnitude_t<T>`; `sigma` as `Σ|z|²`; complex branch; `conj` on `v` in both apply functions
- `qr.hpp` — `conj(tau)` in Q extraction, `conj(v)` in the solve's inner product
- `lq.hpp` — conjugated row + `Hᴴ` on the factor side, plain `tau` on extraction
- `hessenberg.hpp`, `eigenvalue_symmetric.hpp` — explicit complex rejection
- `tests/unit/operation/test_qr.cpp` — complex QR/LQ/householder cases
- `tests/unit/compile_fail/{hessenberg,eigenvalue_symmetric}_complex.cpp`

## Test results

| target | gcc build | gcc test | clang build | clang test |
|---|---|---|---|---|
| `op_test_qr` | OK | PASS (373 assertions / 13 cases) | OK | PASS |
| full unit suite | OK | **160/160** | OK | **160/160** |

## Test plan
- [ ] Fast CI passes (gcc + clang)
- [ ] Promote to ready when satisfied

## Note on merge order

Independent of #360 — different files, no conflicts. Whichever merges second should switch its `Relates to #353` to `Resolves`, or #353 can be closed by hand once both land.

Relates to #353

Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added support for complex-valued Householder reflections, QR factorization, QR solving, and LQ factorization.
  * Improved complex matrix reconstruction and unitary-factor handling across square, tall, and wide matrices.

* **Bug Fixes**
  * Complex-valued operations now correctly apply conjugation where required.

* **Compatibility**
  * Complex inputs to symmetric eigenvalue and Hessenberg factorization are now rejected with a clear compile-time diagnostic.

### CODE REVIEW

File: include/mtl/operation/householder.hpp [hunk 1]
```diff
@@ -36,57 +49,102 @@ std::pair<vec::dense_vector<T>, T> householder(const vec::dense_vector<T>& x) {
...
+    mag_t sigma = mag_t(0);
     for (size_type i = 1; i < n; ++i) {
         const T xs = x(i) / scale;
-        sigma += xs * xs;
+        sigma += functor::scalar::real<T>::apply(T(xs * conj_t::apply(xs)));
     }
 
-    if (sigma == math::zero<T>())
+    if (sigma == mag_t(0))
         return {v, math::zero<T>()};   // x is already along e_1
```
[the remark this answers is not in the document]
Comment (Ravenwater) [reply include/mtl/operation/householder.hpp]: **Confirmed and fixed** in 861d48e. Verified before changing anything:

```
negligible tail, complex x0: tau=0+0i  H*x[0]=1+1i   <- contract says imag==0
R(0,0)=1+1i   R(1,1)=3+0.5i
```

after:

```
negligible tail, complex x0: tau=1.70711-0.707107i  H*x[0]=-1.41421+0i
R(0,0)=-1.41421+0i   R(1,1)=-3.04138+0i
```

Your diagnosis is exactly right, including the mechanism (the early return sits above the `if constexpr`) and the reference behaviour (`zlarfg` returns `tau = 0` only when the tail vanishes **and** alpha is real).

I want to add the part that makes it worth fixing rather than documenting away, because "beta is real" can read as cosmetic. **It is the factor's diagonal.** A complex `R(k,k)` is not merely unconventional — the QR is still valid, Q is still unitary, `Q*R` still reproduces `A`, which is precisely why nothing failed. But a real triangular diagonal is what LAPACK guarantees, and any consumer that assumes it (a condition estimate reading `|R(k,k)|` off the real part, a rank test, anything comparing diagonal entries) silently gets the wrong thing. The defect is invisible from inside the factorization and only visible from outside it, which is the worst place for it to hide.

Your suggested fix is what I took, with one adjustment. I kept both cases in a single `if (sigma == mag_t(0))` with the `if constexpr` nested inside, rather than restructuring the early return, so the real path's control flow is untouched — the real-input bit-exactness invariant this PR asserts is easier to keep true than to re-establish.

`functor::scalar::imag` does exist (`mtl/functor/scalar/imag.hpp`); it is already used by `ldlt` and `cholesky` for the same kind of test, so no new functor was needed.

Falling through turns out to need no special case, which is worth recording since it is not obvious: with `sigma == 0` the complex branch gets `nrm = |x0|`, `beta = -sign(Re x0)*|x0|`, and `d = x0 - beta` of magnitude `2|x0|` — better conditioned than the general case, not worse. `x0` cannot be zero there, because `scale > 0` and `sigma == 0` together force `|x0| == scale`.

Tests added: the negligible-tail case, its complement (a **real** leading entry must still take the shortcut, `tau == 0` exactly, so the fix does not over-trigger), `n == 1`, `n == 0`, and a separate case asserting the complex `R` diagonal is real — the consequence rather than the mechanism.

File: include/mtl/operation/eigenvalue_symmetric.hpp [hunk 2]
```diff
@@ -28,6 +29,23 @@ namespace mtl {
...
+    // householder() and apply_householder_* became complex-capable in #353, so
+    // this would now COMPILE for complex where it previously did not. Reject it
+    // explicitly rather than let a fix elsewhere open a silent-wrong-answer
+    // path here; the Hermitian reduction (real subdiagonal, phase accumulation)
+    // is its own piece of work.
+    static_assert(!is_complex_v<typename M::value_type>,
+        "eigenvalue_symmetric_generic applies H*A*H, a similarity transform only for REAL Householder "
+        "reflectors, which are Hermitian. Complex element types need the unitary "
+        "reduction A -> H*A*H^H and are not supported yet (#353).");
+
```
[the remark this answers is not in the document]
Comment (Ravenwater) [reply include/mtl/operation/eigenvalue_symmetric.hpp]: **Confirmed and fixed** in 861d48e. You are right, and the error is worse than your report states — so let me be plain about it.

**The guard was in the wrong function.** `eigenvalue_symmetric_generic` does not call `householder` at all. `eigen_symmetric` is the one that applies `H*A*H`, and it was the one left open. I did not merely fail to protect a second entry point; I protected a function that this PR never put at risk.

And the compile-failure test made it look fine. It called `eigenvalue_symmetric_generic`, hit the assert that was sitting there, matched the expected regex, and went green — **covering none of the hazard it claimed to close.** That is the exact failure mode this harness was added for in #358, committed inside the harness itself. The negative control now discriminates properly: removing the guard from `eigen_symmetric` fails only that test, which the previous version could not do no matter what I broke.

One correction to the severity, since it changes how urgent this is:

**`eigen_symmetric` still does not compile for complex.** Measured on both sides:

```
this branch : 15 errors      main : 25 errors
```

Making `householder()` complex-capable removed some of the errors that were incidentally holding it shut, but not enough to open it. So **no silent wrong answer was reachable**, on this branch or before it — the "different eigenvalues, no diagnostic" outcome we are both worried about could not actually occur yet.

That does not make the guard pointless, and I have kept it, for two reasons worth distinguishing:

1. **A clean diagnostic instead of a cascade.** Fifteen template errors pointing into the middle of a tridiagonalization is not a usable answer to "why can't I do this?"
2. **A pin.** The incidental errors are exactly the kind that a future type-generalization removes one at a time without noticing. The guard plus its test means the day the last one goes, the build fails loudly rather than the routine quietly returning a non-similar matrix.

I also **removed** the assert from `eigenvalue_symmetric_generic`. Whether that routine handles complex is a genuine question, but it is a pre-existing one that this PR does not touch, and leaving a `#353` assert there would have implied I had checked it. I have not.

`hessenberg_factor` was correct as placed — it does contain its own reduction — and its test was already targeting the right function.

**Comment by Ravenwater** [comment]
## Review round 1 resolved — both findings real, one of them my own harness failing

| # | finding | verdict | action |
|---|---|---|---|
| 1 | `sigma == 0` shortcut breaks the real-beta guarantee for complex | **confirmed, Major** | shortcut now taken only when exact |
| 2 | `eigen_symmetric` not covered by the complex guard | **confirmed — guard was in the wrong function** | moved; test retargeted |
| 3 | Missing Householder edge cases (nitpick) | accepted | 4 cases + an R-diagonal case |

**Finding 1** is silent by construction, which is what makes it worth fixing: Q stays unitary and `Q*R` still reproduces `A`, so nothing in the existing tests could see it. The consequence lives in the factor's diagonal:

```
before   R(0,0)=1+1i        R(1,1)=3+0.5i
after    R(0,0)=-1.41421    R(1,1)=-3.04138
```

A real triangular diagonal is what LAPACK guarantees and what a condition estimate or rank test would assume. Invisible from inside the factorization; wrong from outside it.

**Finding 2 is the one worth dwelling on**, because it is mine and it is embarrassing in a specific way. The guard was in `eigenvalue_symmetric_generic`, which **does not call `householder` at all**. The function that applies `H*A*H` is `eigen_symmetric`, and it was open. The compile-failure test targeted the guarded-but-irrelevant function, matched its regex, and passed — **proving nothing about the hazard it was written to close**.

That is precisely the failure mode the compile-failure harness was introduced for in #358: a test that is green because it is aimed at the wrong thing. I built the mechanism this week, cited it twice as the answer to exactly this problem, and then walked into it. The lesson is not "write compile-failure tests" — it is that a test only means something once you have watched it fail for the reason you intend, and I ran that negative control on the *guard* without checking that the guard was on the *function that mattered*.

Severity correction, stated because it cuts the other way: **`eigen_symmetric` still does not compile for complex** — 15 errors on this branch against 25 on `main`. My change removed some incidental errors, not enough to open it. No silent wrong answer was ever reachable. The guard buys a clean diagnostic and a pin against the day the remaining incidental errors are cleaned up, not a bug fix.

I also removed the assert from `eigenvalue_symmetric_generic`. Whether it handles complex is a real question, but a pre-existing one this PR does not touch, and leaving a `#353` assert there would have implied I had checked. I have not.

Unchanged after both fixes:

| check | result |
|---|---|
| complex QR — 4×4, 6×3, 3×6, 5×5 | `‖QᴴQ−I‖ ≤ 7.8e-16`, `‖QR−A‖ ≤ 7.0e-16` |
| complex LQ — 4×4, 3×6, 6×3 | `‖QᴴQ−I‖ ≤ 4.4e-16`, `‖LQ−A‖ ≤ 6.9e-16` |
| complex `qr_solve` | 3.3e-16 |
| **real path vs `main`** | **bit-identical** (hex-float digest) |

`op_test_qr` 391 assertions / 15 cases. gcc and clang: full build clean, **160/160**.
